File: src/hover/latexCitation.ts
import { TextDocumentPositionParams } from 'vscode-languageserver';
import { Document } from '../document';
import { BibtexFormatter } from '../formatting/bibtex';
import { Language } from '../language';
import { generateBibliography } from '../metadata/bibtexEntry';
import * as range from '../range';
import { HoverProvider } from './provider';

export const LatexCitationHoverProvider: HoverProvider = {
  execute: async ({ document, relatedDocuments, params }) => {
    const name = getEntryName(document, params);
    if (name === undefined) {
      return undefined;
    }

    const entry = getEntry(relatedDocuments, name);
    if (entry === undefined) {
      return undefined;
    }

    const formatter = new BibtexFormatter(true, 4, -1);
    const contents = generateBibliography(formatter.format(entry));
    return contents === undefined ? undefined : { contents };
  },
};

function getEntry(relatedDocuments: Document[], name: string) {
  for (const { tree } of relatedDocuments) {
    if (tree.language === Language.Bibtex) {
      for (const entry of tree.entries) {
        if (entry.name !== undefined && entry.name.text === name) {
          return entry;
        }
      }
    }
  }
  return undefined;
}

function getEntryName(
  { tree }: Document,
  { position }: TextDocumentPositionParams,
): string | undefined {
  switch (tree.language) {
    case Language.Latex:
      for (const { command, name } of tree.citations) {
        if (range.contains(command.range, position)) {
          return name.text;
        }
      }
      break;
    case Language.Bibtex:
      for (const { name } of tree.entries) {
        if (name !== undefined && range.contains(name.range, position)) {
          return name.text;
        }
      }
      break;
  }
  return undefined;
}

File: src/hover/latex_citation.rs
use crate::citeproc::render_citation;
use texlab_protocol::RangeExt;
use crate::syntax::*;
use crate::workspace::*;
use futures_boxed::boxed;
use log::warn;
use lsp_types::*;

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct LatexCitationHoverProvider;

impl FeatureProvider for LatexCitationHoverProvider {
    type Params = TextDocumentPositionParams;
    type Output = Option<Hover>;

    #[boxed]
    async fn execute<'a>(
        &'a self,
        request: &'a FeatureRequest<TextDocumentPositionParams>,
    ) -> Option<Hover> {
        let (tree, entry) = Self::get_entry(request)?;
        if entry.is_comment() {
            None
        } else {
            let key = entry.key.as_ref().unwrap().text();
            match render_citation(&tree, key) {
                Some(markdown) => Some(Hover {
                    contents: HoverContents::Markup(markdown),
                    range: None,
                }),
                None => {
                    warn!("Failed to render entry: {}", key);
                    None
                }
            }
        }
    }
}

impl LatexCitationHoverProvider {
    fn get_entry(
        request: &FeatureRequest<TextDocumentPositionParams>,
    ) -> Option<(&BibtexSyntaxTree, &BibtexEntry)> {
        let key = Self::get_key(request)?;
        for document in request.related_documents() {
            if let SyntaxTree::Bibtex(tree) = &document.tree {
                for entry in tree.entries() {
                    if let Some(current_key) = &entry.key {
                        if current_key.text() == key {
                            return Some((tree, entry));
                        }
                    }
                }
            }
        }
        None
    }

    fn get_key(request: &FeatureRequest<TextDocumentPositionParams>) -> Option<&str> {
        match &request.document().tree {
            SyntaxTree::Latex(tree) => tree
                .citations
                .iter()
                .flat_map(LatexCitation::keys)
                .find(|citation| citation.range().contains(request.params.position))
                .map(LatexToken::text),
            SyntaxTree::Bibtex(tree) => {
                for entry in tree.entries() {
                    if let Some(key) = &entry.key {
                        if key.range().contains(request.params.position) {
                            return Some(key.text());
                        }
                    }
                }
                None
            }
        }
    }
}
